compiler/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -401,19 +402,30 @@ impl<'f> PerFunctionContext<'f> {
         &mut self,
         block_id: BasicBlockId,
         references: &mut Block,
-        mut instruction: InstructionId,
+        instruction: InstructionId,
     ) {
-        // If the instruction was simplified and optimized out of the program we shouldn't analyze
-        // it. Analyzing it could make tracking aliases less accurate if it is e.g. an ArrayGet
+        // If the instruction was simplified and optimized out of the program we shouldn't analyze it.
+        // Analyzing it could make tracking aliases less accurate if it is e.g. an ArrayGet
         // call that used to hold references but has since been optimized out to a known result.
-        if let Some(new_id) = self.inserter.push_instruction(instruction, block_id) {
-            instruction = new_id;
-        } else {
-            return;
-        }
+        // However, if we don't analyze it, then it may be a MakeArray replacing an ArraySet containing references,
+        // and we need to mark those references as used to keep their stores alive.
```

Comment:
If we are only interested in the `MakeArray` case when it's been simplified, I wonder if we could use an early return instead of the many `!simplified`. Maybe something like this:

```rust
        let instr = &self.inserter.function.dfg[instruction];
        if simplified && !matches!(instr, Instruction::MakeArray { .. }) {
            return;
        }

        match instr {
```

(though I'd rename `instruction` to `instruction_id`, then `inst` above to `instruction`)